components/automate-cli/cmd/chef-automate/ha_node_operations_utils.go

```diff
@@ -321,6 +349,176 @@ func (nu *NodeUtilsImpl) executeShellCommand(command, path string) error {
 	}, path)
 }
 
+// Stop services on node
+func (nu *NodeUtilsImpl) stopServicesOnNode(ip, nodeType, deploymentType string, infra *AutomateHAInfraDetails) error {
+
+	if nodeType == OPENSEARCH {
+		err := nu.excludeOpenSearchNode(ip, infra)
+		if err != nil {
+			return err
+		}
+		nu.writer.Println(fmt.Sprintf("OpenSearch node %s excluded from the cluster.", ip))
+	}
+
+	// If deployment type is AWS, then we don't need to stop services on node
+	if deploymentType == AWS_MODE {
+		return nil
+	}
+
+	var cmd string
+	switch nodeType {
+	case AUTOMATE:
+		cmd = STOP_FE_SERVICES_CMD
+	case CHEF_SERVER:
+		cmd = STOP_FE_SERVICES_CMD
+	case POSTGRESQL:
+		cmd = STOP_BE_SERVICES_CMD
+	case OPENSEARCH:
+		cmd = STOP_BE_SERVICES_CMD
+	default:
+		return errors.New("Invalid node type")
+	}
+
+	sshUtil := nu.cmdUtil.GetSshUtil()
+
+	sshUtil.setSSHConfig(&SSHConfig{
+		sshUser:    infra.Outputs.SSHUser.Value,
+		sshPort:    infra.Outputs.SSHPort.Value,
+		sshKeyFile: infra.Outputs.SSHKeyFile.Value,
+		hostIP:     ip,
+	})
+
+	_, err := sshUtil.connectAndExecuteCommandOnRemote(cmd, false)
+	if err != nil {
+		return err
+	}
+
+	nu.writer.Println(fmt.Sprintf("Services stopped on the node %s.", ip))
+
+	return nil
+}
+
+func (nu *NodeUtilsImpl) excludeOpenSearchNode(ipToDelete string, infra *AutomateHAInfraDetails) error {
+
+	automateIp := infra.Outputs.AutomatePrivateIps.Value[0]
+
+	existingIps, err := nu.checkExistingExcludedOSNodes(automateIp, infra)
+	if err != nil {
+		return err
+	}
+
+	if strings.Contains(existingIps, ipToDelete) {
+		return errors.New("Node is already excluded from the cluster")
+	}
+
+	if existingIps != "" {
+		ipToDelete = existingIps + "," + ipToDelete
+	}
+
+	nodeMap := &NodeTypeAndCmd{
+		Frontend:   &Cmd{CmdInputs: &CmdInputs{NodeType: false}},
+		Automate:   createCmdInputs(automateIp, fmt.Sprintf(EXCLUDE_OPENSEARCH_NODE_REQUEST, ipToDelete)),
+		ChefServer: &Cmd{CmdInputs: &CmdInputs{NodeType: false}},
+		Postgresql: &Cmd{CmdInputs: &CmdInputs{NodeType: false}},
+		Opensearch: &Cmd{CmdInputs: &CmdInputs{NodeType: false}},
+		Infra:      infra,
+	}
+
+	_, err = nu.cmdUtil.ExecuteWithNodeMap(nodeMap)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (nu *NodeUtilsImpl) checkExistingExcludedOSNodes(automateIp string, infra *AutomateHAInfraDetails) (string, error) {
+	nodeMap := &NodeTypeAndCmd{
+		Frontend: &Cmd{CmdInputs: &CmdInputs{NodeType: false}},
+		Automate: &Cmd{CmdInputs: &CmdInputs{
+			Cmd:                      GET_OPENSEARCH_CLUSTER_SETTINGS,
+			NodeIps:                  []string{automateIp},
+			Single:                   true,
```

Comment:
is this needed if we are passing nodeips array?

Reply:
done